File: c/calculate.c
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>

// Statistics for a particular station
typedef struct {
    char *city;
    size_t city_len;
    uint32_t hash;
    int count;
    int sum;
    int max;
    int min;
} Station;

// Add a single measurement to this station
void station_add_measurement(Station *st, int mnum) {
    st->count++;
    st->sum += mnum;
    if (mnum < st->min) {
        st->min = mnum;
    }
    if (mnum > st->max) {
        st->max = mnum;
    }
}

void station_add_station(Station *st, Station *station) {
    st->count += station->count;
    st->sum += station->sum;
    if (station->min < st->min) {
        st->min = station->min;
    }
    if (station->max > st->max) {
        st->max = station->max;
    }
}

int city_compare(char *ci, size_t ci_len, char *cj, size_t cj_len) {
    size_t min_len = ci_len;
    if (cj_len < min_len) {
        min_len = cj_len;
    }

    int cmp = memcmp(ci, cj, min_len);

    if (cmp == 0) {
        return ci_len - cj_len;
    } else {
        return cmp;
    }
}

int station_compare(const void *i, const void *j) {
    Station *si = *(Station**)i;
    Station *sj = *(Station**)j;

    // > 0 means i bigger than j
    // < 0 means j bigger than i
    // sort nulls to the end means that nulls need to be bigger
    if (si == NULL && sj == NULL) {
        return 0;
    }
    if (si == NULL) {
        return INT_MAX;
    }
    if (sj == NULL) {
        return INT_MIN;
    }

    return city_compare(si->city, si->city_len, sj->city, sj->city_len);
}

const size_t HashTableInitialSize = 1 << 18; // Must be a power of 2

// A collection of station statistics
typedef struct {
    Station **table; // Hash table of stations
    size_t length;
    uint32_t count;
} Stations;

void stations_init(Stations *s) {
    s->count = 0;
    s->length = HashTableInitialSize;
    s->table = (Station**)calloc(s->length, sizeof(Station*));
}

void stations_zero(Stations *s) {
    for (size_t i = 0; i < s->length; i++) {
        Station *st = s->table[i];
        if (st == NULL) {
            continue;
        }
        free(st);
        s->table[i] = NULL;
    }
}
    
void stations_free(Stations *s) {
    free(s->table);
}

// Sort stations by city
// Warning, this destroys the hashtable
void stations_sort(Stations *s) {
    qsort(s->table, s->length, sizeof(Station*), station_compare);
}

// Get a station given a hash and key value
// If the station does not exist, return nil and index where it would go
uint32_t stations_get(Stations *s, uint32_t hash, char *city, size_t city_len, Station **station) {
    // Because size is a power of 2, we can just mask off the higher bits
    // This is faster than using modulus
    uint32_t index = hash & (s->length-1);

    // Forward probe until you find a matching station or an empty entry
    Station *entry = s->table[index];
    while (entry != NULL && city_compare(city, city_len, entry->city, entry->city_len) != 0) {
        index++;
        if (index >= s->length) {
            index = 0; // Loop around to beginning
        }
        entry = s->table[index];
    }
    *station = entry;
    return index;
}

// Update a station value - this value should already exist
void stations_update(Stations *s, Station *value) {
    Station *entry;
    uint32_t index = stations_get(s, value->hash, value->city, value->city_len, &entry);
    s->table[index] = value;
}

// Add a new station to the set - this station should not already exist
// The index should be returned from stations_get()
// Warning: You may need to free the station
void stations_append(Stations *s, uint32_t index, Station *st) {
    s->table[index] = st;
    s->count++;

    // Make sure that the hash table is twice as big as the number of entries
    if (s->count > s->length / 2) {
        size_t oldlength = s->length;
        Station **oldtable = s->table; 
        s->length *= 2;
        s->table = (Station**)calloc(s->length, sizeof(Station*));
        for (size_t i = 0; i < oldlength; i++) {
            Station *station = oldtable[i];
            if (station == NULL) {
                continue;
            }
            stations_update(s, station);
        }
    }
}

// Combine the statistics of several sets of stations
void stations_combine(Stations *s, Stations *others) {
    for (size_t i = 0; i < others->length; i++) {
        Station *other = others->table[i];
        if (other == NULL) {
            continue;
        }
        Station *existing;
        uint32_t index = stations_get(s, other->hash, other->city, other->city_len, &existing);
        if (existing == NULL) {
            stations_append(s, index, other);
        } else {
            station_add_station(existing, other);
        }
    }
}

typedef struct {
    char *chunk;
    size_t chunk_len;
} Chunk;

// Divide data into a set of chunks
// Ensure that the chunks are on newline boundaries
size_t chunk_data(char *data, size_t data_len, Chunk *chunks, size_t numchunks) {
    size_t chunksize = data_len / numchunks;
    size_t start = 0;
    size_t end = start + chunksize - 1;
    size_t index = 0;
    while (start < data_len) {
        if (end >= data_len) {
            end = data_len - 1;
        }
        // Look forward for the next newline
        while (end < data_len && data[end] != '\n') {
            end++;
        }
        chunks[index].chunk = data + start;
        chunks[index].chunk_len = end - start;
        index++;

        start = end + 1;
        end = start + chunksize + 1;
    }
    return index;
}


const uint32_t offset32 = 2166136261;
const uint32_t prime32  = 16777619;

// Parse the next line of the input
// This should be allocation free
// This expects well formatted input for performance reasons
size_t parse_line(char *data, size_t pos, char **city_r, size_t *city_len_r, uint32_t *hash_r, int *number_r) {
    size_t start = pos;

    // Parse the city name
    //   Advance to the position of the semicolon
    //   Calculate the FNV hash at the same time
    uint32_t hash = offset32;
    char value = data[pos];
    while (value != ';') {
        hash = (hash ^ value) * prime32;
        pos++;
        value = data[pos];
    }

    *city_r = data + start;
    *city_len_r = pos - start;
    *hash_r = hash;

    // Skip over the semicolon
    pos++;

    // Look for the negative sign
    int sign = 1;
    if (data[pos] == '-') {
        sign = -1;
        pos++;
    }

    // Parse the temperature whole number part
    //   Advance to the position of the dot
    int whole = 0;
    value = data[pos];
    while (value != '.') {
        whole *= 10;
        whole += value - '0';
        pos++;
        value = data[pos];
    }

    // Skip over the dot
    pos++;

    // Add in the decimal value and the sign
    //   Use the fact that there is only one decimal precision to represent as an int
    //   This allows for integer arithmetic instead of floating points
    *number_r = sign * ((whole * 10) + data[pos]-'0');

    // Skip over the decimal value and the newline
    pos += 2;

    return pos;
}

void parse_chunk(char *data, size_t data_len, Stations *stations) {
    char *city;
    size_t city_len;
    uint32_t hash;
    int mnum;
    size_t pos = 0;
    while (pos < data_len) {
        pos = parse_line(data, pos, &city, &city_len, &hash, &mnum);

        Station *st = NULL;
        uint32_t index = stations_get(stations, hash, city, city_len, &st);
        if (st == NULL) {
            st = (Station*)malloc(sizeof(Station));
            st->city = city;
            st->city_len = city_len;
            st->hash = hash;
            st->count = 1;
            st->sum = mnum;
            st->max = mnum;
            st->min = mnum;
            stations_append(stations, index, st);
        } else {
            station_add_measurement(st, mnum);
        }
    }
}

void format_output(Stations *totals) {
    stations_sort(totals);
    printf("{");
    fflush(stdout);
    for (size_t i = 0; i < totals->count; i++) {
        Station *st = totals->table[i];
        size_t wrote = write(1, st->city, st->city_len);
        if (wrote != st->city_len) {
            perror("error writing stdout");
            exit(1);
        }
        float minimum = (float)(st->min) / 10.0;
        float mean = (float)(st->sum) / (float)(st->count) / 10.0;
        float maximum = (float)(st->max) / 10.0;
        printf("=%.1f/%.1f/%.1f", minimum, mean, maximum);
        if (i < totals->count - 1) {
            printf(", ");
        }
        fflush(stdout);
    }
    printf("}\n");
}

int main(int argc, char *argv[]) {
    const char *filename = "measurements.txt";

    if (argc == 2) {
        filename = argv[1];
    }

    int file = open(filename, O_RDONLY);
    if (file < 0) {
        perror("error opening file");
        return 1;
    }

    struct stat fileinfo;
    int err = fstat(file, &fileinfo);
    if (err < 0) {
        perror("error getting file stats");
        return 1;
    }

    size_t filesize = fileinfo.st_size;

    char *data = mmap(NULL, filesize, PROT_READ, MAP_SHARED, file, 0);
    if (data == MAP_FAILED) {
        perror("error mmaping file");
        return 1;
    }

    Stations stations;
    stations_init(&stations);
    parse_chunk(data, filesize, &stations);

    format_output(&stations);

    stations_zero(&stations);
    stations_free(&stations);

    close(file);
    munmap(data, filesize);

    return 0;
}

File: go/calculate.go
package main

import (
	"bytes"
	"fmt"
	"log"
	"os"
	"runtime"
	"runtime/debug"
	"sort"
	"strings"
	"syscall"
)

// Statistics for a particular station
type Station struct {
	city  []byte
	hash  uint32 // hash of the city
	count int
	sum   int
	max   int
	min   int
}

// Add a single measurement to this station
func (st *Station) AddMeasurement(mnum int) {
	st.count++
	st.sum += mnum
	if mnum < st.min {
		st.min = mnum
	}
	if mnum > st.max {
		st.max = mnum
	}
}

// Add the statistics to this station
func (st *Station) AddStation(station *Station) {
	st.count += station.count
	st.sum += station.sum
	if station.max > st.max {
		st.max = station.max
	}
	if station.min < st.min {
		st.min = station.min
	}
}

const HashTableInitialSize = 1 << 18 // Must be a power of 2

// A collection of station statistics
type Stations struct {
	table []*Station // Hash table of stations
	list  []*Station // Unique list of stations
}

func NewStations() *Stations {
	return &Stations{
		table: make([]*Station, HashTableInitialSize),
		list:  make([]*Station, 0),
	}
}

// The number of unique stations
func (s *Stations) Count() int {
	return len(s.list)
}

// Sort unique stations by city
func (s *Stations) Sort() {
	sort.Slice(s.list, func(i, j int) bool {
		return bytes.Compare(s.list[i].city, s.list[j].city) < 0
	})
}

// All unique stations
func (s *Stations) All() []*Station {
	return s.list
}

// var HashTableCollisions int = 0
// var HashTableExpansions int = 0

// Get a station given a hash and key value
// If the station does not exist, return nil and index where it would go
func (s *Stations) Get(hash uint32, city []byte) (int, *Station) {
	// Because size is a power of 2, we can just mask off the higher bits
	// This is faster than using modulus
	index := int(hash & uint32(len(s.table)-1))
	//index := int(hash % uint32(len(ht.table)))

	// Forward probe until you find a matching station or an empty entry
	entry := s.table[index]
	for entry != nil && !bytes.Equal(city, entry.city) {
		// HashTableCollisions++
		index++
		if index >= len(s.table) {
			index = 0 // Loop around to beginning
		}
		entry = s.table[index]
	}
	return index, entry
}

// Update a station value - this value should already exist
func (s *Stations) Update(value *Station) {
	index, _ := s.Get(value.hash, value.city)
	s.table[index] = value
}

// Add a new station to the set - this station should not already exist
// The index should be returned from Get()
func (s *Stations) Append(index int, st *Station) {
	s.table[index] = st
	s.list = append(s.list, st)

	// Make sure that the hash table is twice as big as the number of entries
	if len(s.list) > len(s.table)/2 {
		// HashTableExpansions++
		s.table = make([]*Station, len(s.table)*2)
		for _, station := range s.list {
			s.Update(station)
		}
	}
}

// Combine the statistics of several sets of stations
func (s *Stations) Combine(others *Stations) {
	for _, other := range others.All() {
		index, existing := s.Get(other.hash, other.city)
		if existing == nil {
			s.Append(index, other)
		} else {
			existing.AddStation(other)
		}
	}
}

// Divide data into a set of chunks
// Ensure that the chunks are on newline boundaries
func chunkData(data []byte, numchunks int) [][]byte {
	chunksize := len(data) / numchunks

	chunks := make([][]byte, 0)
	start := 0
	end := start + chunksize - 1
	for start < len(data) {
		if end >= len(data) {
			end = len(data) - 1
		}
		// Look forward for the next newline
		for end < len(data) && data[end] != '\n' {
			end++
		}
		chunk := data[start:end]
		chunks = append(chunks, chunk)

		start = end + 1
		end = start + chunksize - 1
	}
	return chunks
}

// Constants for FNV Hashing
// https://github.com/golang/go/blob/master/src/hash/fnv/fnv.go
const (
	offset32 = 2166136261
	prime32  = 16777619
)

// Parse the next line of the input
// This should be allocation free
// This expects well formatted input for performance reasons
func parseLine(data []byte, pos int) ([]byte, uint32, int, int) {
	start := pos

	// Parse the city name
	//   Advance to the position of the semicolon
	//   Calculate the FNV hash at the same time
	var hash uint32 = offset32
	value := data[pos]
	for value != ';' {
		hash = (hash ^ uint32(value)) * prime32
		pos++
		value = data[pos]
	}

	city := data[start:pos]

	// Skip over the semicolon
	pos++

	// Look for the negative sign
	sign := 1
	if data[pos] == '-' {
		sign = -1
		pos++
	}

	// Parse the temperature whole number part
	//   Advance to the position of the dot
	var whole byte = 0
	value = data[pos]
	for value != '.' {
		whole *= 10
		whole += value - '0'
		pos++
		value = data[pos]
	}

	// Skip over the dot
	pos++

	// Add in the decimal value and the sign
	//   Use the fact that there is only one decimal precision to represent as an int
	//   This allows for integer arithmetic instead of floating points
	number := sign * ((int(whole) * 10) + int(data[pos]-'0'))

	// Skip over the decimal value and the newline
	pos += 2

	return city, hash, number, pos
}

// Parse a chunk of data and return all the statistics on a result channel
func parseChunk(data []byte, result chan<- *Stations) {
	stations := NewStations()
	var city []byte
	var hash uint32
	var mnum int
	var pos int = 0
	for pos < len(data) {
		city, hash, mnum, pos = parseLine(data, pos)

		index, st := stations.Get(hash, city)
		if st == nil {
			st = &Station{city: city, hash: hash, count: 1, sum: mnum, max: mnum, min: mnum}
			stations.Append(index, st)
		} else {
			st.AddMeasurement(mnum)
		}
	}
	result <- stations
}

// Format the output as expected by the description
func formatOutput(totals *Stations) string {
	totals.Sort()
	output := make([]string, totals.Count())
	for i, st := range totals.All() {
		key := st.city
		minimum := float32(st.min) / 10.0
		mean := float32(st.sum) / float32(st.count) / 10.0
		maximum := float32(st.max) / 10.0
		output[i] = fmt.Sprintf("%s=%.1f/%.1f/%.1f", key, minimum, mean, maximum)
	}
	return fmt.Sprintf("{%s}", strings.Join(output, ", "))
}

func run() error {
	debug.SetGCPercent(-1) // disable garbage collection
	//debug.SetMemoryLimit(math.MaxInt64)

	// Enable cpu profiling
	// cpuf, _ := os.Create("cpu.prof")
	// defer cpuf.Close()
	// pprof.StartCPUProfile(cpuf)
	// defer pprof.StopCPUProfile()

	// Open and mmap the file
	file, err := os.Open(os.Args[1])
	if err != nil {
		return fmt.Errorf("error opening file: %w", err)
	}
	defer file.Close()

	fileinfo, err := file.Stat()
	if err != nil {
		return fmt.Errorf("error getting file stats: %w", err)
	}

	filesize := fileinfo.Size()

	data, err := syscall.Mmap(int(file.Fd()), 0, int(filesize), syscall.PROT_READ, syscall.MAP_SHARED)
	if err != nil {
		return fmt.Errorf("error mmaping file: %w", err)
	}
	defer syscall.Munmap(data)

	// Chunk the data
	numchunks := runtime.NumCPU()
	// numchunks = 1
	chunks := chunkData(data, numchunks)

	// Parse the chunks as separate goroutines
	results := make(chan *Stations)
	for _, chunk := range chunks {
		go parseChunk(chunk, results)
	}

	// Combine the results
	totals := NewStations()
	for range chunks {
		totals.Combine(<-results)
	}

	// Take a memory snapshot
	// memf, _ := os.Create("mem.prof")
	// defer memf.Close()
	// pprof.WriteHeapProfile(memf)

	// fmt.Fprintln(os.Stderr, "Collisions:", HashTableCollisions, "Expansions:", HashTableExpansions)

	// Show the output
	output := formatOutput(totals)
	fmt.Println(output)

	return nil
}

func main() {
	err := run()
	if err != nil {
		log.Fatal(err)
	}
}
